### PR TITLE
feat: new ory theme using elements

### PR DESCRIPTION
**Pull request by Benehiko**
## Related Issue or Design Document

Implements https://github.com/ory-corp/cloud/issues/2015

## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md) and signed the CLA.
- [ ] I have referenced an issue containing the design document if my change introduces a new feature.
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security vulnerability, 
      I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation within the code base (if appropriate).

### CODE REVIEW

File: src/routes/login.ts [hunk 1]
```diff
@@ -72,5 +99,9 @@ export const registerLoginRoute: RouteRegistrator = (
   app,
   createHelpers = defaultConfig,
 ) => {
-  app.get("/login", createLoginRoute(createHelpers))
+  app.get(
+    "/login",
+    //requireNoAuth(createHelpers),
```
Comment (jonas-jonas) [inline src/routes/login.ts:104]: Remove?

File: src/routes/registration.ts [hunk 2]
```diff
@@ -45,8 +51,27 @@ export const createRegistrationRoute: RouteCreator =
...
+            groups: ["webauthn"],
+            attributes: ["button"],
+            withoutDefaultAttributes: true,
+            withoutDefaultGroup: true,
+          })
+            .filter(({ attributes }) => isUiNodeInputAttributes(attributes))
+            .map(({ attributes }) => {
+              return (attributes as UiNodeInputAttributes).onclick
+            })
+            .filter((c) => c !== undefined),
```
Comment (jonas-jonas) [inline src/routes/registration.ts:66]:
```suggestion
            .filter((onClickAction) => !!onClickAction),
```

File: src/routes/registration.ts [hunk 3]
```diff
@@ -45,8 +51,27 @@ export const createRegistrationRoute: RouteCreator =
...
+          card: UserAuthCard({
+            title: "Register an account",
+            flow: flow as SelfServiceFlow,
```
Comment (jonas-jonas) [inline src/routes/registration.ts:69]: I think we should either add a check that `flow` actually satisfies the type or adjust it in elements. Either way, this typecast is not great, IMO.

File: src/routes/registration.ts [hunk 4]
```diff
@@ -45,8 +51,27 @@ export const createRegistrationRoute: RouteCreator =
...
         // Render the data using a view (e.g. Jade Template):
         res.render("registration", {
-          ...flow,
-          signInUrl: initLoginUrl,
+          nodes: flow.ui.nodes,
+          webAuthnHandler: filterNodesByGroups({
+            nodes: flow.ui.nodes,
+            groups: ["webauthn"],
+            attributes: ["button"],
+            withoutDefaultAttributes: true,
```
Comment (jonas-jonas) [inline src/routes/registration.ts:59]: Not really related to this PR, but negative flags/params/attributes always seem a bit "smelly" to me. Instead, I'd prefer `withDefaultAttributes` and handle its omission accordingly (e.g. determine a good default value/behavior for it). WDYT? Can we still change that?

File: src/routes/sessions.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        await sdk
+          .createSelfServiceLogoutFlowUrlForBrowsers(req.header("cookie"))
+          .catch(() => ({ data: { logout_url: "" } }))
+      ).data.logout_url || ""
+
+    res.render("session", {
+      layout: "welcome",
+      traits: {
+        ...session?.identity.traits,
+        ...{ created_at: session?.identity.created_at || "" },
```
Comment (jonas-jonas) [inline src/routes/sessions.ts:27]: This doesn't need to be spread, right?

File: src/routes/sessions.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        ...session?.authentication_methods?.reduce<any>(
```
Comment (jonas-jonas) [inline src/routes/sessions.ts:28]: This reducer is really hard to read. If I read it correctly, this would produce the following entries in `traits`:

```js
{
  //... other traits
  "authentication_method: 4 Jun 2017 00:00:00 PDT": "password",
  //... other traits
}
```

Do we need that key to use it as a look-up? That seems _really_ hacky and awkward to me.

File: src/routes/settings.ts [hunk 7]
```diff
@@ -32,11 +43,99 @@ export const createSettingsRoute: RouteCreator =
...
+            groups: "webauthn",
+            attributes: "button",
+            withoutDefaultAttributes: true,
+            withoutDefaultGroup: true,
+          })
+            .filter(({ attributes }) => isUiNodeInputAttributes(attributes))
+            .map(({ attributes }) => {
+              return (attributes as UiNodeInputAttributes).onclick
+            })
+            .filter((c) => c !== undefined),
```
Comment (jonas-jonas) [inline src/routes/settings.ts:132]:
```suggestion
            .filter((onClickAction) => !!onClickAction),
```

File: src/routes/welcome.ts [hunk 8]
```diff
@@ -22,12 +23,40 @@ export const createWelcomeRoute: RouteCreator =
...
-      hasSession: Boolean(session),
-      logoutUrl,
+      layout: "welcome",
+      nav: navigationMenu(session, logoutUrl, "welcome"),
+      concepts: [
+        CardGradient({
+          heading: "Getting Started",
+          content:
+            "Jump start your project and complete the quickstart tutorial to get a broader overview of Ory Cloud.",
+          action: "/",
```
Comment (jonas-jonas) [inline src/routes/welcome.ts:33]: We still need the correct URLs here, right?

File: src/routes/welcome.ts [hunk 9]
```diff
@@ -22,12 +23,40 @@ export const createWelcomeRoute: RouteCreator =
...
+            "Define Text here Jump start your project and complete the quickstart tutorial to get a broader",
+          action: "/",
+        }),
+        CardGradient({
+          heading: "Custom UI",
+          content:
+            "Implementing these pages in your language and framework of choice is straightforward using our SDKs.",
+          action: "/",
+        }),
+      ].join("\n"),
```
Comment (jonas-jonas) [inline src/routes/welcome.ts:59]: How does this work? Does `CardGradient` just return a string?

File: views/layouts/auth.hbs [hunk 10]
```diff
@@ -0,0 +1,29 @@
+<html lang='en'>
+  <head>
+    <meta content="width=device-width, initial-scale=1" name="viewport" />
+    <link rel='icon' type='image/png' href='favico.png' />
+    <meta charset='utf-8' />
+    <title>{{projectName}}</title>
+    {{! Themes }}
+    <link rel="stylesheet" href="style.css" />
```
Comment (jonas-jonas) [inline views/layouts/auth.hbs:8]: Do we know the absolute path to the `style.css` here? I always find myself debugging these relative imports if anything changes on the server side, so I prefer absolute imports to make it clear, which file is meant.

Comment (jonas-jonas) [reply views/layouts/auth.hbs:8]: These same imports are also used on the other files, maybe we could extract the `link` or the whole header to a fragment?

File: views/partials/ui_node_script.hbs [hunk 11]
```diff
@@ -6,4 +6,5 @@
   crossorigin="{{attributes.crossorigin}}"
   {{#if attributes.async}}async{{/if}}
   data-testid="node/script/{{attributes.id}}"
+  async="true"
```
Comment (jonas-jonas) [inline views/partials/ui_node_script.hbs:9]: How does this work with `{{#if attributes.async}}async{{/if}}` a few lines above?

File: views/session.hbs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+<script type="application/javascript">
+  // Get the current breakpoint
+  const getBreakpoint = function () {
+    return window.getComputedStyle(document.body, ':before').content.replace(/\"/g, '');
+  };
+
+  const setSessionCodeBoxCollapse = function () {
+    breakpoint = getBreakpoint();
+    const codeBoxInput = document.getElementsByClassName("session-code-box")[0].getElementsByTagName("input")[0]
```
Comment (jonas-jonas) [inline views/session.hbs:36]: Could we move this into the `CodeBox` element? I think this should be possible with pure CSS as well, if we refactor the CodeBox a bit.

File: views/settings.hbs [hunk 13]
```diff
@@ -1,80 +1,49 @@
...
-  <div class="card">
-    <div class="card-action">
-      <a class="typography-link typography-h2" href="welcome">Back</a>
-    </div>
-  </div>
-</div>
+  {{> ui_nodes nodes=nodes groups="webauthn" attributes="text/javascript" withoutDefaultGroup=true}}
+
+    <script type="text/javascript">
+    eval(document.getElementsByName("webauthn_register_trigger").forEach((v) =>
```
Comment (jonas-jonas) [inline views/settings.hbs:43]: Does this need to be `eval`'d? It should just render as text on the HTML page and be executed, right?

**Review by jonas-jonas — CHANGES_REQUESTED** [review]
A few things.

Also, a lot of the files don't have new lines at the end. I'll check if our new formatting tasks are present in this repo and fix it in a separate PR.

File: src/routes/registration.ts (hunk 4 above)
Comment (Benehiko) [reply src/routes/registration.ts:59]: Yeah I understand what you mean. We cannot change this now since this is in a library which had to get tested for this. I think in future we can update it, but for now the priority is to get this released.
https://github.com/ory/integrations/blob/main/src/ui/index.ts#L142-L149

File: src/routes/registration.ts (hunk 2 above)
Comment (Benehiko) [reply src/routes/registration.ts:66]: nice :)

File: src/routes/registration.ts (hunk 3 above)
Comment (Benehiko) [reply src/routes/registration.ts:69]: This is casting it to ory elements type which is the same underlying type given by the SDK.

So e.g. here flow can be SelfServiceRegistrationFlow but ory/elements expects SelfServiceFlow which is a type defined as 
```ts
type SelfServiceFlow = SelfServiceRegistrationFlow | SelfServiceLoginFlow ...
```

I don't see this breaking somewhere. We could move this type to the SDK directly at some point and have it used everywhere instead of this workaround.

File: src/routes/sessions.ts (hunk 5 above)
Comment (Benehiko) [reply src/routes/sessions.ts:27]: Ah yes, you're right. it was just remnants of refactoring :)

File: src/routes/sessions.ts (hunk 6 above)
Comment (Benehiko) [reply src/routes/sessions.ts:28]: this is indeed hacky and was rushed a bit here to get the UI working as I wanted it to.

The keys are just used the UI to display the traits in a "form-like" structure.

Comment (Benehiko) [reply src/routes/sessions.ts:28]: Here is an example

image

File: src/routes/welcome.ts (hunk 8 above)
Comment (Benehiko) [reply src/routes/welcome.ts:33]: yes :)

File: src/index.ts [hunk 14]
```diff
@@ -48,11 +95,12 @@ registerRecoveryRoute(app)
 registerRegistrationRoute(app)
 registerSettingsRoute(app)
 registerVerificationRoute(app)
+registerSessionsRoute(app)
 registerWelcomeRoute(app)
 registerErrorRoute(app)
 
 app.get("/", (req: Request, res: Response) => {
-  res.redirect("welcome", 303)
+  res.redirect(303, "sessions")
```
Comment (alnr) [inline src/index.ts:103]: This should redirect to `welcome`, I think?

File: src/routes/registration.ts (hunk 3 above)
Comment (jonas-jonas) [reply src/routes/registration.ts:69]: Hmm, I see. 

>I don't see this breaking somewhere. 

Famous last words ;)

> `type SelfServiceFlow = SelfServiceRegistrationFlow | SelfServiceLoginFlow ...`

In that case, the typecast should not be necessary at all, right?

File: src/routes/registration.ts (hunk 4 above)
Comment (jonas-jonas) [reply src/routes/registration.ts:59]: Oh, I missed that this is part of an older library. Alright. 

Thought this is part of `elements`...

File: package.json [hunk 15]
```diff
@@ -18,12 +18,15 @@
     "serve": "node lib/index.js",
     "start": "nodemon",
     "test": "exit 0",
-    "prepublishOnly": "npm run build"
+    "format": "prettier --write .",
```
Comment (jonas-jonas) [inline package.json:21]: Please remove this, and use `make format` instead.

**Comment by alnr** [comment]
Bug on the session page:
image

File: views/session.hbs (hunk 12 above)
Comment (Benehiko) [reply views/session.hbs:36]: I don't think you can do this with only css since you cannot change the state (input element is checked) of the collapse based on the screen size in css.

Since how this works is you have a checkbox hidden and a label (icon) linked to the checkbox. Since the checkbox is hidden the label is clickable and sets the state (checked or not checked). CSS can react to an input box state but not change it.

Comment (jonas-jonas) [reply views/session.hbs:36]: >I don't think you can do this with only css since you cannot change the state (input element is checked) of the collapse based on the screen size in css.

I'd have to think about it. I really don't like doing these kinds of things on load, and I think it should be possible to have the interactivity just in CSS. It might be more hacky than this, though. So not sure if that'd be a win. :D 

The `breakpoint` variable should be `const` or `let`. Just noticed.

File: package.json [hunk 16]
```diff
@@ -18,12 +18,15 @@
...
+    "format:check": "prettier --check .",
+    "prepublishOnly": "npm run build",
+    "postinstall": "mkdir -p elements; cd elements; git init; git remote add -f origin https://github.com/ory/elements.git; git pull --depth=1 origin main; git checkout v0.0.1-alpha.0; npm i; npm run initialize; npm run build:markup; cd ..; npm i ./elements/packages/markup"
```
[the remark this answers is not in the document]
Comment (Benehiko) [reply package.json:24]: originally with this setup i could do some pre-setup like setting `git config core.sparsecheckout true`.
But i think cloning like this would also be fine.

File: .gitignore [hunk 17]
```diff
@@ -98,3 +98,10 @@ jspm_packages/
 
 lib
 .idea
+
+pnpm-lock.yaml
```
[the remark this answers is not in the document]
Comment (Benehiko) [reply .gitignore:102]: Yeah i will remove `pnpm-lock` from the repo. I accidentally committed it and forgot to remove it. It's supposed to work seemlessly side-by-side, but I think just sticking with npm would be best.

File: src/routes/registration.ts (hunk 4 above)
Comment (Benehiko) [reply src/routes/registration.ts:59]: all good :)

File: src/routes/registration.ts (hunk 3 above)
Comment (Benehiko) [reply src/routes/registration.ts:69]: you are correct :)
I think i added this long ago when i was still busy with elements...

File: src/routes/welcome.ts (hunk 9 above)
Comment (Benehiko) [reply src/routes/welcome.ts:59]: Yes, and handlebars injects the array of strings into the template as just string which is rendered as HTML.

Comment (jonas-jonas) [reply src/routes/welcome.ts:59]: I see, should `\n` not be `<br />` then?

File: src/index.ts (hunk 14 above)
Comment (Benehiko) [reply src/index.ts:103]: I don't think this should be to welcome since the hosted UI will directly be fed to customers as a "login" and "registration" screen. The welcome page is thus hidden from end-users and only developers launching it from the console will be directly redirected to the welcome page.

File: views/session.hbs (hunk 12 above)
Comment (Benehiko) [reply views/session.hbs:36]: Hmm, yeah i would also need to think about it a bit. But this was the best solution I could come up with for the collapse on breakpoint without breaking the codebox entirely.

File: views/settings.hbs (hunk 13 above)
Comment (Benehiko) [reply views/settings.hbs:43]: you are right, it doesn't need eval. I think this was just a lot of work to get it working with lots of refactoring which ended up keeping eval ;)

File: views/layouts/auth.hbs (hunk 10 above)
Comment (Benehiko) [reply views/layouts/auth.hbs:8]: since this is an express application, the client side part of this will be the browser loading the `style.css` sheet from the server as a relative URL to whatever the browser url.

e.g. https://custom-domain.com will load the `style.css` as https://custom-domain.com/style.css

The browser handles this loading and this isn't a filesystem load like in nextjs or react.

Comment (Benehiko) [reply views/layouts/auth.hbs:8]: Yeah i think we can maybe move this to a partial

File: src/routes/welcome.ts (hunk 9 above)
Comment (Benehiko) [reply src/routes/welcome.ts:59]: no, since this is just a simple array that is concatenating it. we can simply do a `.join("")` as well. but the html won't look so pretty on render :)

File: views/layouts/auth.hbs (hunk 10 above)
Comment (jonas-jonas) [reply views/layouts/auth.hbs:8]: Yes, I know^^

But `style.css` is a relative import and those can be hard to debug, because they're relative to the file including them. That's why I prefer absolute imports, e.g. `/some/static/path/styles.css`.

File: views/session.hbs (hunk 12 above)
Comment (jonas-jonas) [reply views/session.hbs:36]: Sounds good, for now, then.

File: src/routes/welcome.ts (hunk 9 above)
Comment (jonas-jonas) [reply src/routes/welcome.ts:59]: Ah, alright. Kinda weird to see a `\n` in combination with HTML, though ^^. But LGTM then.

File: views/layouts/auth.hbs (hunk 10 above)
Comment (Benehiko) [reply views/layouts/auth.hbs:8]: but you cannot do absolute imports here since this is an http call back to itself on the root path

Comment (jonas-jonas) [reply views/layouts/auth.hbs:8]: Eh, absolute paths don't work when the UI is hosted on a path, that's not the root of a domain (e.g. `http://custom-domain.com/auth/login`). Damn, that's annoying. But doesn't matter I guess. LGTM.

Comment (Benehiko) [reply views/layouts/auth.hbs:8]: the express server is explicitly serving the file from the root path.
https://github.com/ory/kratos-selfservice-ui-node/pull/189/files#diff-1f5f84e6c3633404ab2cfd17752c70ae17ff2f973991206a8d2495e3f708674fL7

 we can change this to whatever we want including the `/auth/login` :)

File: src/routes/sessions.ts [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+          (methods, method, i) => {
+            methods[`authentication method used`] = `${method.method} (${
+              method.completed_at && new Date(method.completed_at).toUTCString()
+            })`
+            return methods
+          },
+          {},
+        ),
+        "authentiction level":
```
Comment (jonas-jonas) [inline src/routes/sessions.ts:54]: typo `authentication level`

**Review by jonas-jonas — APPROVED** [review]
LGTM. Ship it 🎉

File: src/index.ts (hunk 14 above)
Comment (aeneasr) [reply src/index.ts:50]: This will have a massive impact on the Ory Proxy and Ory Tunnel and potentially the CNAME system. Please create an issue explaining what has changed so we can work on dealing with this in the Ory CLI.

**Comment by aeneasr** [comment]
Great job! I tried covering all the flows I know. There are a few style issues I found, and some problems with error messages, but otherwise this looks very advanced!

## auth template Style differences

1. Buttons are too high and the corners not as rounded
2. Borders have a darker gray
3. Spacings are off below the logo, sub-headline, sign in button, horizontal bar
4. Horizontal bar has wrong shade of gray
5. bottom bar is missing (on purpose?)

Screenshot 2022-10-04 at 18 08 32

## 404 error not showing

Call any URL (e.g. http://localhost:4455/dgsfgsdfgw435) and you get a blank page:

Screenshot 2022-10-04 at 18 13 55

## Welcome page

1. Docs buttons have gradient always, not just on hover
2. Docs buttons are sized higher and wider

Screenshot 2022-10-04 at 18 14 55

## Session data

1. Not using code font
2. Hide contents looks a bit weird when it's empty. Also the text doesn't change to "show contents". Maybe rename to "Toggle content"?

Screenshot 2022-10-04 at 18 19 41
Screenshot 2022-10-04 at 18 20 17

## Error pages

They're missing the logo, and the message is not formatted correctly.

Screenshot 2022-10-04 at 18 21 39

## Settings

1. Just ename to "Account settings"
2. Potentially add short description
3. Buttons are too wide
4. Spacing in header nav bar is not correct
5. Nav links do not work (the page does not jump)
6. Nav links they show entries for things that do not exist in the form (e.g. 2fa backup codes)

Screenshot 2022-10-04 at 18 23 36

## Recovery

1. Rename `Already have an account? Sign in` -> `Remember your credentials? Sign in`
2. Similar to login also button sizes and margins

Screenshot 2022-10-04 at 18 29 09

## Form errors

They are shown as global errors instead of as errors in context of the field. This will become problematic if there are more than one errors (e.g. minimum length is 7) as users won't know which error this references to.

Screenshot 2022-10-04 at 18 33 40

**Comment by Benehiko** [comment]
Fixes related to UI are also in [elements](https://github.com/ory/elements/pull/17):
- [X] Gradient card size and hover effect
- [X] Nav heading space fixed
- [X] Auth Card spacing between elements
- [X] Settings page Nav buttons
- [X] Settings page button sizes
- [X] Settings page description
- [X] Error card image and text
- [X] Session page Mono font
- [X] Codebox collapse and text
- [X] 404 page working
- [ ] Auth card rounding
- [ ] Divider shade of grey 
- [ ] Auth card border shade of grey
- [ ] OAuth2 flow text on auth card
- [ ] Auth Card error messages per field instead of global
- [ ] base url fixes for stylesheets (e.g. on /randompath/anotherpath/ should load styles from the base path / or /anything)

File: src/index.ts (hunk 14 above)
Comment (Benehiko) [reply src/index.ts]: yeah I reverted this to `welcome` again. I did test it with the ory cli proxy and couldn't find a problem. But there might be some edge case I am missing. We can evaluate it later

**Comment by Benehiko** [comment]
all kratos e2e tests pass against the latest commit

**Comment by Benehiko** [comment]
image

e2e tests are run against https://github.com/ory/kratos/pull/2708
